File: internal/advisor/scoring.go
package advisor

import (
	"fmt"
	"math"
	"sort"
	"strings"
	"time"

	"protonsage/internal/core"
)

const (
	freshDays      = 90
	recentDays     = 365
	historicalDays = 730
)

type SimilarityResult = core.SimilarityResult
type RankedReport = core.RankedReport

// RecencyScore weights fresh reports strongly and drops reports older than a year sharply.
func RecencyScore(reportTime, now time.Time) float64 {
	if reportTime.IsZero() {
		return 0
	}
	if now.IsZero() {
		now = time.Now().UTC()
	}
	ageDays := now.Sub(reportTime).Hours() / 24
	if ageDays <= 0 {
		return 1
	}

	switch {
	case ageDays <= 30:
		return 1
	case ageDays <= freshDays:
		return roundScore(0.95 - ((ageDays-30)/(freshDays-30))*0.10)
	case ageDays < recentDays:
		return roundScore(0.75 - ((ageDays-freshDays)/(recentDays-freshDays))*0.35)
	case ageDays < historicalDays:
		return roundScore(0.24 - ((ageDays-recentDays)/(historicalDays-recentDays))*0.12)
	default:
		return 0.05
	}
}

// FreshnessLabel classifies reports for UI/CLI stale warnings.
func FreshnessLabel(reportTime, now time.Time) string {
	age := ReportAgeDays(reportTime, now)
	switch {
	case age < freshDays:
		return core.FreshnessFresh
	case age < recentDays:
		return core.FreshnessRecent
	case age < historicalDays:
		return core.FreshnessStale
	default:
		return core.FreshnessHistorical
	}
}

// ReportAgeDays returns whole report age in days, clamped to zero for future timestamps.
func ReportAgeDays(reportTime, now time.Time) int {
	if reportTime.IsZero() {
		return 0
	}
	if now.IsZero() {
		now = time.Now().UTC()
	}
	age := int(now.Sub(reportTime).Hours() / 24)
	if age < 0 {
		return 0
	}
	return age
}

// SystemSimilarity compares normalized local and report system profiles. Unknown fields are explanatory but neutral.
func SystemSimilarity(user, report core.NormalizedSystemProfile) SimilarityResult {
	fields := []similarityField{
		{name: "GPU vendor", user: user.GPUVendor, report: report.GPUVendor, weight: 0.28},
		{name: "GPU model", user: user.GPUModel, report: report.GPUModel, weight: 0.12, partial: true},
		{name: "GPU driver", user: user.GPUDriver, report: report.GPUDriver, weight: 0.10},
		{name: "distro family", user: user.DistroFamily, report: report.DistroFamily, weight: 0.15},
		{name: "kernel", user: user.Kernel, report: report.Kernel, weight: 0.10},
		{name: "RAM bucket", user: user.RAMBucket, report: report.RAMBucket, weight: 0.08},
		{name: "session type", user: user.SessionType, report: report.SessionType, weight: 0.07},
		{name: "CPU vendor", user: user.CPUVendor, report: report.CPUVendor, weight: 0.05},
		{name: "CPU class", user: user.CPUClass, report: report.CPUClass, weight: 0.05},
	}

	var matchedWeight, comparableWeight float64
	result := SimilarityResult{}
	for _, field := range fields {
		userValue := normalizeComparableValue(field.user)
		reportValue := normalizeComparableValue(field.report)
		if userValue == "" || reportValue == "" {
			result.Unknowns = append(result.Unknowns, unknownFieldExplanation(field.name, userValue, reportValue))
			continue
		}
		comparableWeight += field.weight
		matchScore := fieldMatchScore(userValue, reportValue, field.partial)
		if matchScore > 0 {
			matchedWeight += field.weight * matchScore
			result.Matches = append(result.Matches, fmt.Sprintf("%s %s", strings.ToLower(field.name), reportValue))
		} else {
			result.Mismatches = append(result.Mismatches, fmt.Sprintf("%s user=%s report=%s", strings.ToLower(field.name), userValue, reportValue))
		}
	}

	if comparableWeight == 0 {
		result.Score = 0.5
		result.Summary = "No comparable system fields were available."
		return result
	}
	result.Score = roundScore(matchedWeight / comparableWeight)
	result.Summary = similaritySummary(result)
	return result
}

// RankReports returns reports ordered by deterministic recency-first advisor score.
func RankReports(reports []core.Report, profile core.SystemProfile, now time.Time) []RankedReport {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	userProfile := profile.Normalized
	if normalizedProfileIsZero(userProfile) {
		userProfile = core.NormalizeSystemProfile(profile)
	}

	ranked := make([]RankedReport, 0, len(reports))
	for _, report := range reports {
		recency := RecencyScore(report.Timestamp, now)
		similarity := SystemSimilarity(userProfile, normalizedReportProfile(report))
		quality := ReportQualityScore(report)
		freshness := FreshnessLabel(report.Timestamp, now)
		item := RankedReport{
			Report:           report,
			Score:            roundScore(recency*0.72 + similarity.Score*0.20 + quality*0.08),
			RecencyScore:     recency,
			SystemSimilarity: similarity.Score,
			QualityScore:     quality,
			Freshness:        freshness,
			AgeDays:          ReportAgeDays(report.Timestamp, now),
			Similarity:       similarity,
			Reasons:          rankReasons(report, freshness, recency, similarity, quality, now),
		}
		ranked = append(ranked, item)
	}

	sort.SliceStable(ranked, func(i, j int) bool {
		if ranked[i].Score == ranked[j].Score {
			if ranked[i].Report.Timestamp.Equal(ranked[j].Report.Timestamp) {
				return reportStableID(ranked[i].Report) < reportStableID(ranked[j].Report)
			}
			return ranked[i].Report.Timestamp.After(ranked[j].Report.Timestamp)
		}
		return ranked[i].Score > ranked[j].Score
	})
	return ranked
}

// ReportQualityScore gives a small boost to positive, concrete reports without overpowering recency.
func ReportQualityScore(report core.Report) float64 {
	score := 0.45
	switch strings.ToLower(strings.TrimSpace(report.Rating)) {
	case "platinum":
		score = 0.95
	case "gold":
		score = 0.85
	case "silver":
		score = 0.65
	case "bronze":
		score = 0.45
	case "borked":
		score = 0.20
	}
	if strings.TrimSpace(report.LaunchOptions) != "" || textMentionsConcreteTweak(report.Notes) {
		score += 0.08
	}
	if strings.TrimSpace(report.Verdict) != "" {
		score += 0.03
	}
	return roundScore(clamp(score, 0, 1))
}

type similarityField struct {
	name    string
	user    string
	report  string
	weight  float64
	partial bool
}

func fieldMatchScore(user, report string, partial bool) float64 {
	if user == report {
		return 1
	}
	if partial && len(user) >= 4 && len(report) >= 4 && (strings.Contains(user, report) || strings.Contains(report, user)) {
		return 0.7
	}
	return 0
}

func normalizeComparableValue(value string) string {
	value = strings.TrimSpace(strings.ToLower(value))
	if value == "" || value == "unknown" {
		return ""
	}
	return value
}

func unknownFieldExplanation(name, userValue, reportValue string) string {
	name = strings.ToLower(name)
	switch {
	case userValue == "" && reportValue == "":
		return name + " unknown for user and report"
	case userValue == "":
		return name + " unknown for user"
	default:
		return name + " unknown for report"
	}
}

func similaritySummary(result SimilarityResult) string {
	parts := []string{}
	if len(result.Matches) > 0 {
		parts = append(parts, "matches "+strings.Join(limitStrings(result.Matches, 3), ", "))
	}
	if len(result.Mismatches) > 0 {
		parts = append(parts, "differs on "+strings.Join(limitStrings(result.Mismatches, 3), ", "))
	}
	if len(result.Unknowns) > 0 {
		parts = append(parts, "unknown "+strings.Join(limitStrings(result.Unknowns, 2), ", "))
	}
	if len(parts) == 0 {
		return "No comparable system fields were available."
	}
	return strings.Join(parts, "; ") + "."
}

func rankReasons(report core.Report, freshness string, recency float64, similarity SimilarityResult, quality float64, now time.Time) []string {
	reasons := []string{
		fmt.Sprintf("%s report (%d days old, recency %.2f)", freshness, ReportAgeDays(report.Timestamp, now), recency),
		fmt.Sprintf("system similarity %.2f: %s", similarity.Score, similarity.Summary),
		fmt.Sprintf("quality %.2f from rating/verdict/concrete tweak signals", quality),
	}
	if strings.TrimSpace(report.LaunchOptions) != "" {
		reasons = append(reasons, "explicit launch options field present")
	}
	if freshness == core.FreshnessStale || freshness == core.FreshnessHistorical {
		reasons = append(reasons, "treat as historical context, not current truth")
	}
	return reasons
}

func normalizedReportProfile(report core.Report) core.NormalizedSystemProfile {
	profile := core.NormalizeSystemInfoMap(report.SystemInfo)
	if report.SystemInfo == nil {
		return profile
	}
	apply := func(target *string, keys ...string) {
		for _, key := range keys {
			if value := strings.TrimSpace(report.SystemInfo[key]); value != "" {
				*target = strings.ToLower(value)
				return
			}
		}
		for _, key := range keys {
			for actualKey, value := range report.SystemInfo {
				if strings.EqualFold(actualKey, key) && strings.TrimSpace(value) != "" {
					*target = strings.ToLower(strings.TrimSpace(value))
					return
				}
			}
		}
	}
	apply(&profile.GPUVendor, "normalized.gpuVendor")
	apply(&profile.GPUModel, "normalized.gpuModel")
	apply(&profile.GPUDriver, "normalized.gpuDriver")
	apply(&profile.CPUVendor, "normalized.cpuVendor")
	apply(&profile.CPUClass, "normalized.cpuClass")
	apply(&profile.RAMBucket, "normalized.ramBucket")
	apply(&profile.DistroFamily, "normalized.distroFamily")
	apply(&profile.Kernel, "normalized.kernel")
	apply(&profile.SessionType, "normalized.sessionType")
	return profile
}

func normalizedProfileIsZero(profile core.NormalizedSystemProfile) bool {
	return profile == (core.NormalizedSystemProfile{})
}

func reportStableID(report core.Report) string {
	if report.SourceReportID != "" {
		return report.SourceReportID
	}
	if report.ID != 0 {
		return fmt.Sprintf("%d", report.ID)
	}
	return fmt.Sprintf("%d:%s", report.AppID, report.Timestamp.UTC().Format(time.RFC3339))
}

func limitStrings(values []string, limit int) []string {
	if len(values) <= limit {
		return values
	}
	out := append([]string{}, values[:limit]...)
	out = append(out, fmt.Sprintf("+%d more", len(values)-limit))
	return out
}

func roundScore(value float64) float64 {
	return math.Round(value*1000) / 1000
}

func clamp(value, min, max float64) float64 {
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

File: src/advisor/scoring.cpp
#include "advisor.h"
#include <QtMath>
#include <algorithm>
#include <cmath>

namespace ProtonSage {

// ── Internal helpers ─────────────────────────────────────────────────

namespace {

constexpr int kFreshDays      = 90;
constexpr int kRecentDays     = 365;
constexpr int kHistoricalDays = 730;

double roundScore(double value) {
    return std::round(value * 1000.0) / 1000.0;
}

double clamp(double value, double min, double max) {
    return std::clamp(value, min, max);
}

// Normalize a comparable value: lowercase, trimmed, "unknown" → empty.
QString normalizeComparableValue(const QString& value) {
    QString v = value.trimmed().toLower();
    if (v.isEmpty() || v == QStringLiteral("unknown"))
        return {};
    return v;
}

// Score for a single field comparison (exact or partial match).
double fieldMatchScore(const QString& user, const QString& report, bool partial) {
    if (user == report)
        return 1.0;
    if (partial && user.size() >= 4 && report.size() >= 4 &&
        (user.contains(report) || report.contains(user)))
        return 0.7;
    return 0.0;
}

QString unknownFieldExplanation(const QString& name, const QString& userValue, const QString& reportValue) {
    if (userValue.isEmpty() && reportValue.isEmpty())
        return name + QStringLiteral(" unknown for user and report");
    if (userValue.isEmpty())
        return name + QStringLiteral(" unknown for user");
    return name + QStringLiteral(" unknown for report");
}

QString similaritySummary(const SimilarityResult& result) {
    QStringList parts;

    auto limit = [](const QStringList& values, int n) -> QStringList {
        if (values.size() <= n) return values;
        QStringList out = values.mid(0, n);
        out << QString::number(values.size() - n) + QStringLiteral(" more");
        return out;
    };

    if (!result.matches.isEmpty())
        parts << QStringLiteral("matches ") + limit(result.matches, 3).join(QStringLiteral(", "));
    if (!result.mismatches.isEmpty())
        parts << QStringLiteral("differs on ") + limit(result.mismatches, 3).join(QStringLiteral(", "));
    if (!result.unknowns.isEmpty())
        parts << QStringLiteral("unknown ") + limit(result.unknowns, 2).join(QStringLiteral(", "));

    if (parts.isEmpty())
        return QStringLiteral("No comparable system fields were available.");

    return parts.join(QStringLiteral("; ")) + QStringLiteral(".");
}

// Build a SystemProfile from a Report's systemInfo map,
// preferring keys with "normalized." prefix (case-insensitive fallback).
SystemProfile normalizedReportProfile(const Report& report) {
    SystemProfile profile;

    // Helper: try exact keys first, then case-insensitive.
    auto apply = [&](QString& target, const QStringList& keys) {
        for (const auto& key : keys) {
            auto it = report.systemInfo.constFind(key);
            if (it != report.systemInfo.constEnd()) {
                QString v = it.value().trimmed().toLower();
                if (!v.isEmpty()) { target = v; return; }
            }
        }
        // Case-insensitive fallback
        for (const auto& key : keys) {
            for (auto it = report.systemInfo.constBegin(); it != report.systemInfo.constEnd(); ++it) {
                if (it.key().compare(key, Qt::CaseInsensitive) == 0) {
                    QString v = it.value().trimmed().toLower();
                    if (!v.isEmpty()) { target = v; return; }
                }
            }
        }
    };

    apply(profile.normGpuVendor,    QStringList{QStringLiteral("normalized.gpuVendor")});
    apply(profile.normGpuModel,     QStringList{QStringLiteral("normalized.gpuModel")});
    apply(profile.normGpuDriver,    QStringList{QStringLiteral("normalized.gpuDriver")});
    apply(profile.normCpuVendor,    QStringList{QStringLiteral("normalized.cpuVendor")});
    apply(profile.normCpuClass,     QStringList{QStringLiteral("normalized.cpuClass")});
    apply(profile.normRamBucket,    QStringList{QStringLiteral("normalized.ramBucket")});
    apply(profile.normDistroFamily, QStringList{QStringLiteral("normalized.distroFamily")});
    apply(profile.normKernel,       QStringList{QStringLiteral("normalized.kernel")});
    apply(profile.normSessionType,  QStringList{QStringLiteral("normalized.sessionType")});

    return profile;
}

bool normalizedProfileIsZero(const SystemProfile& p) {
    return p.normGpuVendor.isEmpty()  && p.normGpuModel.isEmpty() && p.normGpuDriver.isEmpty()
        && p.normCpuVendor.isEmpty()  && p.normCpuClass.isEmpty()
        && p.normRamBucket.isEmpty()
        && p.normDistroFamily.isEmpty()
        && p.normKernel.isEmpty()
        && p.normSessionType.isEmpty();
}

QString reportStableID(const Report& report) {
    if (!report.sourceReportId.isEmpty())
        return report.sourceReportId;
    if (report.id != 0)
        return QString::number(report.id);
    return QString::number(report.appId) + QStringLiteral(":") + report.timestamp.toUTC().toString(Qt::ISODate);
}

struct SimilarityField {
    QString name;
    QString user;
    QString report;
    double weight = 0;
    bool partial = false;
};

} // anonymous namespace

// ── Public API ───────────────────────────────────────────────────────

double recencyScore(const QDateTime& reportTime, const QDateTime& now) {
    if (!reportTime.isValid())
        return 0.0;
    QDateTime n = now.isValid() ? now : QDateTime::currentDateTimeUtc();
    double ageDays = reportTime.msecsTo(n) / 86400000.0;
    if (ageDays <= 0.0)
        return 1.0;

    if (ageDays <= 30.0)
        return 1.0;
    if (ageDays <= kFreshDays)
        return roundScore(0.95 - ((ageDays - 30.0) / (kFreshDays - 30.0)) * 0.10);
    if (ageDays < kRecentDays)
        return roundScore(0.75 - ((ageDays - kFreshDays) / (kRecentDays - kFreshDays)) * 0.35);
    if (ageDays < kHistoricalDays)
        return roundScore(0.24 - ((ageDays - kRecentDays) / (kHistoricalDays - kRecentDays)) * 0.12);
    return 0.05;
}

QString freshnessLabel(const QDateTime& reportTime, const QDateTime& now) {
    int age = reportAgeDays(reportTime, now);
    if (age < kFreshDays)     return Freshness::Fresh;
    if (age < kRecentDays)    return Freshness::Recent;
    if (age < kHistoricalDays) return Freshness::Stale;
    return Freshness::Historical;
}

int reportAgeDays(const QDateTime& reportTime, const QDateTime& now) {
    if (!reportTime.isValid()) return 0;
    QDateTime n = now.isValid() ? now : QDateTime::currentDateTimeUtc();
    int age = static_cast<int>(reportTime.msecsTo(n) / 86400000.0);
    return std::max(0, age);
}

SimilarityResult systemSimilarity(const SystemProfile& user, const SystemProfile& report) {
    QList<SimilarityField> fields = {
        {QStringLiteral("GPU vendor"),    user.normGpuVendor,    report.normGpuVendor,    0.28, false},
        {QStringLiteral("GPU model"),     user.normGpuModel,     report.normGpuModel,     0.12, true},
        {QStringLiteral("GPU driver"),    user.normGpuDriver,    report.normGpuDriver,    0.10, false},
        {QStringLiteral("distro family"), user.normDistroFamily, report.normDistroFamily, 0.15, false},
        {QStringLiteral("kernel"),        user.normKernel,       report.normKernel,       0.10, false},
        {QStringLiteral("RAM bucket"),    user.normRamBucket,    report.normRamBucket,    0.08, false},
        {QStringLiteral("session type"),  user.normSessionType,  report.normSessionType,  0.07, false},
        {QStringLiteral("CPU vendor"),    user.normCpuVendor,     report.normCpuVendor,    0.05, false},
        {QStringLiteral("CPU class"),     user.normCpuClass,      report.normCpuClass,     0.05, false},
    };

    double matchedWeight = 0.0;
    double comparableWeight = 0.0;
    SimilarityResult result;

    for (const auto& field : fields) {
        QString userValue   = normalizeComparableValue(field.user);
        QString reportValue = normalizeComparableValue(field.report);

        if (userValue.isEmpty() || reportValue.isEmpty()) {
            result.unknowns.append(unknownFieldExplanation(field.name, userValue, reportValue));
            continue;
        }

        comparableWeight += field.weight;
        double matchScore = fieldMatchScore(userValue, reportValue, field.partial);
        if (matchScore > 0.0) {
            matchedWeight += field.weight * matchScore;
            result.matches.append(field.name.toLower() + QStringLiteral(" ") + reportValue);
        } else {
            result.mismatches.append(field.name.toLower()
                                     + QStringLiteral(" user=") + userValue
                                     + QStringLiteral(" report=") + reportValue);
        }
    }

    if (comparableWeight == 0.0) {
        result.score = 0.5;
        result.summary = QStringLiteral("No comparable system fields were available.");
        return result;
    }

    result.score = roundScore(matchedWeight / comparableWeight);
    result.summary = similaritySummary(result);
    return result;
}

// ── ReportQualityScore ──────────────────────────────────────────────

namespace {

double reportQualityScore(const Report& report) {
    double score = 0.45;
    QString rating = report.rating.trimmed().toLower();

    if (rating == QStringLiteral("platinum"))  score = 0.95;
    else if (rating == QStringLiteral("gold"))   score = 0.85;
    else if (rating == QStringLiteral("silver")) score = 0.65;
    else if (rating == QStringLiteral("bronze")) score = 0.45;
    else if (rating == QStringLiteral("borked")) score = 0.20;

    // Boost for concrete tweak signals
    if (!report.launchOptions.trimmed().isEmpty())
        score += 0.08;
    else if (report.notes.contains(QStringLiteral("%command%"), Qt::CaseInsensitive))
        score += 0.08;
    else {
        // Check for env assignments or wrapper mentions in notes
        static const QRegularExpression envRe(QStringLiteral("\\b[A-Z_][A-Z0-9_]{2,}="));
        static const QRegularExpression wrapperRe(
            QStringLiteral("\\b(gamemoderun|mangohud|gamescope|prime-run|obs-gamecapture)\\b"),
            QRegularExpression::CaseInsensitiveOption);
        if (report.notes.indexOf(envRe) >= 0 || report.notes.indexOf(wrapperRe) >= 0)
            score += 0.08;
    }

    if (!report.verdict.trimmed().isEmpty())
        score += 0.03;

    return roundScore(clamp(score, 0.0, 1.0));
}

} // namespace

// ── RankReports ─────────────────────────────────────────────────────

QList<RankedReport> rankReports(const QList<Report>& reports,
                                 const SystemProfile& profile,
                                 const QDateTime& now) {
    QDateTime n = now.isValid() ? now : QDateTime::currentDateTimeUtc();
    SystemProfile userProfile = profile;
    Q_UNUSED(userProfile);

    QList<RankedReport> ranked;
    ranked.reserve(reports.size());

    for (const auto& report : reports) {
        double recency   = recencyScore(report.timestamp, n);
        SimilarityResult sim = systemSimilarity(profile, normalizedReportProfile(report));
        double quality   = reportQualityScore(report);
        QString freshness = freshnessLabel(report.timestamp, n);
        int ageDays      = reportAgeDays(report.timestamp, n);

        RankedReport item;
        item.report           = report;
        item.score            = roundScore(recency * 0.72 + sim.score * 0.20 + quality * 0.08);
        item.recencyScore     = recency;
        item.systemSimilarity = sim.score;
        item.qualityScore     = quality;
        item.freshness         = freshness;
        item.ageDays           = ageDays;
        item.similarity        = sim;

        // Build reasons
        QStringList reasons;
        reasons << QString(QStringLiteral("%1 report (%2 days old, recency %3"))
                       .arg(freshness).arg(ageDays).arg(recency, 0, 'f', 3);
        reasons << QString(QStringLiteral("system similarity %1: %2"))
                       .arg(sim.score, 0, 'f', 3).arg(sim.summary);
        reasons << QString(QStringLiteral("quality %1 from rating/verdict/concrete tweak signals"))
                       .arg(quality, 0, 'f', 3);

        if (!report.launchOptions.trimmed().isEmpty())
            reasons << QStringLiteral("explicit launch options field present");

        if (freshness == Freshness::Stale || freshness == Freshness::Historical)
            reasons << QStringLiteral("treat as historical context, not current truth");

        item.reasons = reasons;
        ranked.append(item);
    }

    // Sort: descending score, then descending timestamp, then stable ID
    std::sort(ranked.begin(), ranked.end(),
              [](const RankedReport& a, const RankedReport& b) {
                  if (qFuzzyCompare(a.score, b.score)) {
                      if (a.report.timestamp == b.report.timestamp) {
                          return reportStableID(a.report) < reportStableID(b.report);
                      }
                      return a.report.timestamp > b.report.timestamp;
                  }
                  return a.score > b.score;
              });

    return ranked;
}

} // namespace ProtonSage
